src/relayclient/KnownRelaysManager.ts

```diff
@@ -103,14 +104,55 @@ export default class KnownRelaysManager implements IKnownRelaysManager {
     return origRelays.filter(this.relayFilter)
   }
 
+  splitRange (fromBlock: number, toBlock: number, splits: number): Array<{fromBlock: number, toBlock: number}> {
+    const totalBlocks = toBlock - fromBlock + 1
+    const splitSize = Math.ceil(totalBlocks / splits)
+
+    const ret: Array<{ fromBlock: number, toBlock: number }> = []
+    let b
+    for (b = fromBlock; b < toBlock; b += splitSize) {
+      ret.push({ fromBlock: b, toBlock: Math.min(toBlock, b + splitSize - 1) })
+    }
+    return ret
+  }
+
+  // return events from hub. split requested range into "window parts", to avoid
+  // fetching too many events at once.
+  async getPastEventsForHub (fromBlock: number, toBlock: number): Promise<EventData[]> {
+    let relayEventParts: any[]
+    while (true) {
+      const rangeParts = this.splitRange(fromBlock, toBlock, this.config.relayLookupWindowParts)
+      try {
+        // eslint-disable-next-line @typescript-eslint/promise-function-async
+        const getPastEventsPromises = rangeParts.map(({ fromBlock, toBlock }): Promise<any> =>
+          this.contractInteractor.getPastEventsForHub([], {
+            fromBlock,
+            toBlock
+          }))
+        relayEventParts = await Promise.all(getPastEventsPromises)
+        break
+      } catch (e) {
+        if (e.toString().match(/query returned more than/) != null &&
+          this.config.relayLookupWindowBlocks > this.config.relayLookupWindowParts
+        ) {
+          if (this.config.relayLookupWindowParts >= 16) {
+            throw new Error(`Too many events after splitting by ${this.config.relayLookupWindowParts}`)
+          }
+          this.config.relayLookupWindowParts *= 4
```

Comment:
You are modifying a value in an object that is supposed to be user input and is shared by all components. I suggest KRM to have a member that holds this value if it is to be reused and config to be read-only (something we don't have in code for some reason).

Reply:
the reason to modify it in-place is that future transactions will use same value, and not have to go through the doubling procedure again.
I don't persist it, since I'm not sure its a good value for all future use..
Note sure what you meant by KRM.
it is possible to copy the value out of the config, and into a member of the relayClient